feat: add plugin for ALFA plots

krish8484 opened the pull request:
## Description

Add the main script `multiqc-plugins/modules/ALFA/ALFA.py` and other helper scripts which will complete the job of generating the MultiQC report using `ALFA` plugin


## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have not reduced the existing code coverage

krish8484 commented:
@AngryMaciek I have added the main script and made the necessary changes, you may test it yourself by generating the report using command line instructions form `.github/workflows/build.yml`. There are 4 sub-sections, `Unique-Categories`, `Unique-Biotypes`, `MultipleUnique-Categories` and `MultipleUnique-Biotypes`. I coudn't divide it into 2 subsections with 2 graphs each as apparently each section has a parameter `plot`, so each section must have a graph. 
Let me know what other changes are expected.

AngryMaciek commented on .flake8:
```diff
@@ -0,0 +1,2 @@
+[flake8]
+max-line-length = 89
```
I think that `black` by default has a `88` line.
Docs: https://black.readthedocs.io/en/stable/the_black_code_style.html
> If you’re using Flake8, you can bump max-line-length to 88 and forget about it.

AngryMaciek commented on .github/workflows/build.yml:
```diff
@@ -22,3 +22,6 @@ jobs:
       - name: Install plugin
```
You are installing our whole repository with custom plugins, not just one unnamed "plugin".
*Install plugins* will be OK.

AngryMaciek commented on .github/workflows/build.yml:
```diff
@@ -22,3 +22,6 @@ jobs:
       - name: Install plugin
         run: |
           python setup.py install
+      - name: Plot ALFA plots
```
*Test ALFA plugin*

AngryMaciek commented on .github/workflows/lint.yml:
```diff
@@ -24,3 +24,4 @@ jobs:
         run: |
           black --check setup.py
           black --check modules/ALFA/ALFA.py
+          black --check modules/ALFA/custom_code.py
```
This line is OK but could you please point me to the docs where they say that there has to be an additional file with the hooks?

krish8484 replied:
Ofcourse, two files are not necessary, I was just following `example-plugin`, deleted `custom_code.py` and transferred contents to `ALFA.py`.

AngryMaciek commented on modules/ALFA/ALFA.py:
```diff
@@ -0,0 +1,197 @@
+log = logging.getLogger(__name__)
+
+
+class MultiqcModule(BaseMultiqcModule):
```
Missing class docstring

AngryMaciek commented on modules/ALFA/ALFA.py:
```diff
@@ -0,0 +1,197 @@
+
+
+class MultiqcModule(BaseMultiqcModule):
+    def __init__(self):
```
Missing initializer docstring

AngryMaciek commented on modules/ALFA/__init__.py:
```diff
@@ -0,0 +1,4 @@
+# flake8: noqa
```
Rather than running `flake8` and disabling all errors it would be better not to run the linter on this file at all.

AngryMaciek commented on modules/ALFA/custom_code.py:
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env python
```
This script has a shebang line but `ALFA.py` does not. You need to be consistent.

AngryMaciek reviewed (commented):
Minor code style changes.

AngryMaciek commented:
OK, looks good! I will look in more detail tomorrow and check that everyting is as we expect it.

AngryMaciek commented:
The plots that are already in are good, but we will need a second series of plots for the enrichment.

Please look at the original `ALFA` plots the tool is producing, for example in this [file](../blob/dev/tests/ALFA/Unique/ALFA_plots.Categories.pdf); there are two images. Currently we have the top one, but we will also need the bottom one. I am not sure if we did it previously, but I found the description on how they calculate values for that plot, please take a look at: https://github.com/biocompibens/ALFA#detailed-example

Is seems the `ALFA` tab will have more sections then... I would propose the following structure:
```
ALFA
    Unique-Categories
    Unique-Categories-Enrichment
    Unique-BioTypes
    Unique-BioTypes-Enrichment
    UniqueMultiple-Categories
    UniqueMultiple-Categories-Enrichment
    UniqueMultiple-BioTypes
    UniqueMultiple-BioTypes-Enrichment
```

Actually, when I think of this now, I understand that `Unique` and `UniqueMultiple` are directories names which are being parsed, correct? So, if we run this module on a data just with one directory we would obtain 4 sections, but if there were 4 directories, there would be 16 sections, correct?

krish8484 commented:
Okay, I will go through the `ALFA` example link, understand it and try to add that into the script. The structure is fine.
You have a valid point regarding the number of directories parsed, but that should not be tough to add in the script, I would just need to know at what part will the path be different for example the current data's path is `ALFA/samples/WT_rep1_WT/ALFA/Unique/WT_rep1_WT.ALFA_feature_counts.tsv`, if there was another directory, would it replace `samples` in the path?
If so then the structure should become
```
ALFA
    Sample1- Unique-Categories
    Sample1- Unique-Categories-Enrichment
    Sample1- Unique-BioTypes
    Sample1- Unique-BioTypes-Enrichment
    Sample2- UniqueMultiple-Categories
    Sample2- UniqueMultiple-Categories-Enrichment
    Sample2- UniqueMultiple-BioTypes
    Sample2- UniqueMultiple-BioTypes-Enrichment
```

AngryMaciek commented:
While writting a general plugin we cannot enforce any special structure/depth of the directory tree nor directory/file names except from `*feature_counts.tsv` suffix for the "target file" which will be parsed. Expecting a given directory tree would be too much of a constraint for the users.

The plugin should work as follows:

1. Scan a given target directory, provided in the command line while calling `multiqc` (just as for other modules; let's say `/haha/x`)
2. If you find a match in the filename to a regular expression: `*feature_counts.tsv` then take the parent directory name and make a section in the report out of it (so `/haha/x/a/b/c/LOL.feature_counts.tsv`) would end up in a section `c` in the report.

Unlike you, I did not read on how the pattern search work in `miltiqc` but looking at other internal modules which seem to behave that way this seems reasonable, right?

krish8484 commented:
Now, even if there are additional folders to `Unique` and `UniqueMultiple`, the code would add a new section with the graphs.

krish8484 commented:
In [this](https://github.com/biocompibens/ALFA#detailed-example) example, I did not understand how did we get the `C` graph. I reckon what we already have is the `B` graph. But I did not understand how did we `normalize` `B` to get `C`.

AngryMaciek commented:
In the example, you may notice two colors and a legend: red: bam, green:gtf.
Also, in each of the files which end in `_feature_counts.tsv` there are two columns: *Counts_in_BAM/BedGraph* and *Size_in_genome*. Genome information are usually encoded in GTF format, which links back to the legend.
Therefore I am quite convinced that for each row they calculate `%` alongside both columns and then they make a ratio of these two values. It is interpreted as a ratio of `%` which you observe to the `%` you would expect to observe.

krish8484 commented:
Ok understood the calculations but there is one thing, I don't think it is possible to plot the bargraphs in the negative side of the axis in `MultiQC` like shown in `ALFA` graphs, it all has to be on the positive side. It will still represent the correct information though.

Btw, we can also plot the bars side by side instead of stacking them. I have pushed an example `multiqc_report20.html` (will be deleted later). Which one do we prefer?

AngryMaciek commented:
Nice! I think separate bars look better indeed.

Make sure everything is calculated correctly:
* right now when you hover your mouse over these separate bars the `%` scores in a pop-up table are all zeros
* when you change tabs between Counts/Percantages the bars go back to "stacked" version.

Regarding the bi-directionality, let's investigate more options, as discussed in the meeting, I am looking into possibilities too.

krish8484 commented:
There was a similar issue opened 4 days ago, regarding changing tabs and stacked version reappearing
https://github.com/ewels/MultiQC/issues/1320

AngryMaciek commented:
> There was a similar issue opened 4 days ago, regarding changing tabs and stacked version reappearing
> [ewels/MultiQC#1320](https://github.com/ewels/MultiQC/issues/1320)

OK, then since it is still open let's keep stacked versions for now.

krish8484 commented:
Added the enhancement graphs.
- Currently, I have adjusted the config to represent `non-stacked` version without the tabs to avoid the bugs.
- We can also do the same for the non-enrichment graphs by switching off the counts tab as the counts is already visible in the tool tip and then we can have only the percentages tab with non-stacked bars?
- We can add labels to the axes, please let me know what to add if we want to add them.
- For the enhancement graphs, `log10` has been taken.
- Please do cross check once and confirm that the calculations done is alright.

AngryMaciek commented:
* The `non-stacked` versions look perfect. You are right, let's present only the `%` tab as the actual counts are accesible once you hover your mouse over the plot.
* No, I don't think any labels are neccessary, its good now.
* Could you please use `log2`? People often use `base=2` in the field (https://www.biostars.org/p/303806/) so we should stick to the standard. As I see it: the package `math` also contains a `log2` function so this should be a matter of a few key strokes.
* Your code is very nice: properly formatted and documented. Good job!
* **importantly:** what we calculate here is an *enrichment*, not *enhancement*. Enrichment is a reserved term (like here: https://en.wikipedia.org/wiki/Gene_set_enrichment_analysis). Please change the word everywhere in your code.

krish8484 commented:
We forgot that the tooltip is bugged for `non-stacked`, for instance the counts tab shows percentages incorrectly (`0.0`) and the percentage tab will not be able to show counts correctly.

AngryMaciek commented:
Enrichment plots which are `non-stacked` have a correct tooltip, I do not remember what was the issue with not-Enrichment `non-stacked` versions...

krish8484 commented:
I have pushed two examples, `multiqc_report3.html` and `multiqc_Report5.html`. The former represents non-stacked version of counts and the latter non-stacked version for percentage. The one with the counts have incorrect percentage in the tooltip (we can turn the percentage off though) and the other one does not contain counts in tooltip.

AngryMaciek commented:
Let us revert to `stacked` plots, I would not want to loose information here.

krish8484 commented:
> Let us revert to `stacked` plots, I would not want to loose information here.

`Stacked` for both enrichment and non-enrichment, right?

AngryMaciek commented:
> > Let us revert to `stacked` plots, I would not want to loose information here.
> 
> `Stacked` for both enrichment and non-enrichment, right?

No, I think enrichment plots are fine now, no need to change them.

AngryMaciek reviewed (approved):
The code looks good, ready to merge, there are two very minor issues left:
1) As I see, the report sections are sorted alpahebitally, therefore please adjust the sections names, as you mentioned in this comment https://github.com/zavolanlab/multiqc-plugins/pull/5#issuecomment-712893309
2) In the report: the description should be more informative. We currently have:
> ALFA is an example analysis module used for writing documentation.

Let's change it to the official description from the ALFA github page:
> ALFA provides a global overview of features distribution composing NGS dataset(s). Given a set of aligned reads (BAM files) and an annotation file (GTF format with biotypes), the tool produces plots of the raw and normalized distributions of those reads among genomic categories (stop codon, 5'-UTR, CDS, intergenic, etc.) and biotypes (protein coding genes, miRNA, tRNA, etc.). Whatever the sequencing technique, whatever the organism.

AngryMaciek reviewed (approved):
Perfect!
